### ISSUE TITLE
Certifiability of a promise, related to arbitrary interference

### ISSUE THREAD

**Issue by AnHaechan** [kaist-cp/cs492-concur#168]
I thought
`a promise is certifiable 
if there is a possible execution scenario that it can actually be written,
under some arbitrary interleaving`

But I changed my thought to
`a promise is certifiable
if there is a possible execution scenario that it can actually be written,
under **any** arbitrary interleaving`

For example,

```
{X=0}
r=X
if r!=0 then X=1 
{X=1}
```

I previously thought (X=1) is certifiable because
`X| [0] [666] | 1 |` 
then if r read [666], X=1 can be written.

However, with my new definition of certifiability,
there is an interleaving that does not make promise certifiable:
`X| [0] [0] |1|   `
that is, all interleaving is 0. 
In conclusion, X=1 is not certifiable.

Am I right?
Thanks.

**Comment by totolacky** [comment]
I think you are correct! This actually makes sense, as X=1 should not be hoisted in the given example code.

**Comment by tomtomjhj** [comment]
Right. Promising semantics defines the *consistency* of a thread's state (= (re)certifiability of all promises) as "for all extension of the current memory, there exists an execution of the thread in which all its remaining promises are fulfilled.

**Comment by tomtomjhj** [comment]
More intuitively, that store hoisting doesn't make sense because it changes the semantics of the program in the single-threaded setting.

**Comment by integerblue** [comment]
> Right. Promising semantics defines the _consistency_ of a thread's state (= (re)certifiability of all promises) as "for all extension of the current memory, there exists an execution of the thread in which all its remaining promises are fulfilled".
> More intuitively, that store hoisting doesn't make sense because it changes the semantics of the program in the single-threaded setting.

Actually, I don't understand your two answers. Could you elaborate more on it or translate them into Korean? Thank you.

**Comment by integerblue** [comment]
@tomtomjhj I think I got it. However, then why multi-valued memory is not concerned when checking the consistency of the thread's state? A thread could read any values in the memory timestamp so that the consistency might break.

**Comment by tomtomjhj** [comment]
Consistency check does consider the multi-valued memory. If just suffices to find a scenario in which the promises can be fulfilled.

**Comment by integerblue** [comment]
In #161, @totolacky and I thought that a thread could choose which memory state to use in certification and recertification so that I wondered why, because a thread could use other thread's memory state and fail recertification. 
```
Initially, x = y = 0

Thread 1:
r1 = x
r2 = x
if r1 == r2
  y = 1

Thread 2:
r3 = y
x = r3

Behavior in question: r1 == r2 == r3 == 1
```
We agreed that this could be allowed because even if r1 = x; r3=y; r2=x is the order of execution, r1 and r2 could use its initial memory state 0 so that the recertification succeeds when r3=y is read in thread 2.

**Comment by jeehoonkang** [comment]
@integerblue A thread cannot "choose" the memory state in (re-)certification.  It should certify for **any** possible future memory of the current memory, modeling the arbitrary interference of the other threads.

The reason why your example is allowed is as follows:

- T1 promises to write `y=1` at some timestamp.  It's certified: in any future possible memory, it can still read `x=0` at `r1=x` and `r2=x` and then fulfill `y=1`.

- T2 reads `y=1` and writes `x=1`.

- T1 reads the just-written `x=1` twice and then fulfill `y=1.`

**Comment by integerblue** [comment]
@jeehoonkang I misused the word memory state, but it is rather a message of the memory state. I got the idea! Thanks!
